### PR TITLE
feat: adds support for getting contractType from explorer plugin

### PR DESCRIPTION
**Pull request by sabotagebeats**
### What I did

Integrates getting contract type from explorer plugin

### How I did it

integrates as abstract method

### How to verify it

give it a shot!

h/t @fubuloubu 
```
$ ape console --network ethereum:mainnet:infura

In [1]: c = Contract("0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D")
WARNING: No contract type found for 0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D

In [2]: c
Out[2]: <Address 0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D>

In [3]: exit

$ ape plugins add etherscan

$ ape console --network ethereum:mainnet:infura

In [1]: c = Contract("0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D")

In [2]: c
Out[2]: <BoredApeYachtClub 0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D>

In [4]: c.balanceOf("0xd104Ce1266e5Fa0A31F7CA6dd16DEf1d62308C46")
Out[4]: 1

In [5]: c.ownerOf(4303) == "0xd104Ce1266e5Fa0A31F7CA6dd16DEf1d62308C46"
Out[5]: True
```

### Checklist

- [x] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Comment by fubuloubu** [comment]
```py
$ ape console --network ethereum:mainnet:infura

In [1]: c = Contract("0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D")
WARNING: No contract type found for 0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D

In [2]: c
Out[2]: <Address 0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D>

In [3]: exit

$ ape plugins add etherscan

$ ape console --network ethereum:mainnet:infura

In [1]: c = Contract("0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D")

In [2]: c
Out[2]: <BoredApeYachtClub 0xBC4CA0EdA7647A8aB7C2061c2E118A18a936f13D>

In [4]: c.balanceOf("0xd104Ce1266e5Fa0A31F7CA6dd16DEf1d62308C46")
Out[4]: 1

In [5]: c.ownerOf(4303) == "0xd104Ce1266e5Fa0A31F7CA6dd16DEf1d62308C46"
Out[5]: True
```

File: src/ape/api/explorers.py [hunk 1]
```diff
@@ -36,3 +39,15 @@ def get_transaction_url(self, transaction_hash: str) -> str:
         Returns:
             str
         """
+
+    @abstractmethod
+    def get_contract_type(self, address: AddressType) -> Optional[ContractType]:
+        """
+        Get the contract type for a given address if it has been published in explorer
```
Comment (unparalleled-js) [inline src/ape/api/explorers.py:46]:
```suggestion
        Get the contract type for a given address if it has been published in an explorer.
```

File: src/ape/api/explorers.py [hunk 2]
```diff
@@ -36,3 +39,15 @@ def get_transaction_url(self, transaction_hash: str) -> str:
...
+
+        Args:
+            address (str): The contract address.
+
+        Returns:
+            ContractType if published, else None
```
Comment (unparalleled-js) [inline src/ape/api/explorers.py:52]:
```suggestion
            :class:`~ape.contracts.ContractType` if published, else ``None``.
```

File: src/ape/contracts/base.py [hunk 3]
```diff
@@ -295,27 +304,26 @@ def _Contract(
     #    contract_type = network.contract_cache[address]
 
     # Check explorer API/cache (e.g. publicly published contracts)
-    # TODO: Add ``get_contract_type`` to ``ExplorerAPI``
     # TODO: Store in ``NetworkAPI.contract_cache`` to reduce API calls
-    # explorer = provider.network.explorer
```
Comment (unparalleled-js) [inline src/ape/contracts/base.py:300]: Is this commented out code necessary?

File: src/ape/api/explorers.py (hunk 2 above)
Comment (fubuloubu) [reply src/ape/api/explorers.py:52]: @unparalleled-js we had a question I didn't know offhand, is it acceptable docstring syntax to write a return this way? Thought it was more useful that just saying "returns `Optional[ContractType]` which really doesn't tell you as much about the context shift

**Review by fubuloubu — APPROVED** [review]

File: src/ape/api/explorers.py [hunk 4]
```diff
@@ -36,3 +39,15 @@ def get_transaction_url(self, transaction_hash: str) -> str:
...
+
+    @abstractmethod
+    def get_contract_type(self, address: AddressType) -> Optional[ContractType]:
+        """
+        Get the contract type for a given address if it has been published in an explorer.
+
+        Args:
+            address (str): The contract address.
+
+        Returns:
```
Comment (sabotagebeats) [inline src/ape/api/explorers.py:51]: So this should be 
```
Returns:class:
    `~ape.contracts.ContractType` if published, else ``None``.
```

Comment (sabotagebeats) [reply src/ape/api/explorers.py:51]: no i think it should be 
```
        Returns:
            class:`~ape.contracts.ContractType` if published, else ``None``.
```

**Comment by sabotagebeats** [comment]
@fubuloubu @unparalleled-js  thoughts on how to implement tests for this pr? we have the commands but need to download the etherscan plugin and then run the test with that?

**Comment by fubuloubu** [comment]
> @fubuloubu @unparalleled-js thoughts on how to implement tests for this pr? we have the commands but need to download the etherscan plugin and then run the test with that?

Let's skip it for now. Eventually would like to add a  "mock plugin" that simulates this type of testing in core's test suite so we don't run into these issues.

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**